### DISCUSSION TITLE
Direct dependency with Illuminate\Foundation\Application

### DISCUSSION THREAD

**Discussion by erayaydin** [spatie/laravel-ignition#120] [Q&A]
Hello,

I'm trying to re-structure the Laravel and extend the "Laravel Application" by implementing the `Illuminate\Contracts\Foundation\Application` contract. Trying to register `Laravel Ignition` with my `Laravel Application` concrete but there is a direct dependency in the `Spatie\LaravelIgnition\IgnitionServiceProvider`. I apologize for the submitted issue if it was implemented consciously.

```
TypeError 

Spatie\LaravelIgnition\IgnitionServiceProvider::Spatie\LaravelIgnition\{closure}(): Argument #1 ($app) must be of type Illuminate\Foundation\Application, Foo\Infrastructure\Laravel\Application given, called in ...
```

**Comment by cbaconnier** [comment]
I don't have this issue. How are you registering `Application` ?

Here's an example that works for me.

**Application.php**
```php
namespace App;

class Application extends \Illuminate\Foundation\Application
{
    protected $namespace = 'App\\';
}

```

**app.php**
```php
use App\Application;

$app = new Application(
    $_ENV['APP_BASE_PATH'] ?? dirname(__DIR__)
);

$app->useAppPath($app->basePath('src/App'));
```

Also, I did a package that move classes around to approach a DDD _alike_ architecture shown in [Laravel Beyond Crud](https://laravel-beyond-crud.com/). Maybe you can find some useful information or it could be a good start.

https://github.com/cbaconnier/laravel-mvc-to-ddd

**Reply by erayaydin** [reply]
You are extending concrete Laravel Application (which got extra methods and unnecessary implementations for my architecture). Try to implement `Illuminate\Contracts\Foundation\Application` interface and extend `Illuminate\Container\Container` for re-produce the issue.

> PS: Maybe also need to implement other interfaces if laravel-ignition use (such as `Illuminate\Contracts\Foundation\CachesConfiguration`)

```patch
// src/IgnitionServiceProvider.php
...
-use Illuminate\Foundation\Application;
+use Illuminate\Contracts\Foundation\Application;
```

Thank you for suggestion. I'll read the book and check repository.

I'll also share my Application.

**Reply by erayaydin** [reply]
```php
namespace Foo\Infrastructure\Laravel;

...
use Illuminate\Container\Container;
use Illuminate\Contracts\Foundation\Application as ApplicationContract;
use Illuminate\Contracts\Foundation\CachesConfiguration;
use Illuminate\Contracts\Foundation\CachesRoutes;
use Symfony\Component\HttpKernel\HttpKernelInterface;
...

class Application extends Container implements ApplicationContract, CachesConfiguration, CachesRoutes, HttpKernelInterface
{
  // Custom implementations of methods. I don't want to overwrite Laravel's concrete Application
  ...
}
```
